crates/factor-outbound-http/src/lib.rs

```diff
@@ -38,19 +51,90 @@ impl Factor for OutboundHttpFactor {
-        let allowed_hosts = builders
-            .get_mut::<OutboundNetworkingFactor>()?
-            .allowed_hosts();
+        let outbound_networking = builders.get_mut::<OutboundNetworkingFactor>()?;
+        let allowed_hosts = outbound_networking.allowed_hosts();
+        let component_tls_configs = outbound_networking.component_tls_configs().clone();
         Ok(InstanceState {
-            allowed_hosts,
             wasi_http_ctx: WasiHttpCtx::new(),
+            allowed_hosts,
+            component_tls_configs,
+            request_interceptor: None,
         })
     }
 }
 
 pub struct InstanceState {
-    allowed_hosts: OutboundAllowedHosts,
     wasi_http_ctx: WasiHttpCtx,
+    allowed_hosts: OutboundAllowedHosts,
+    component_tls_configs: ComponentTlsConfigs,
+    request_interceptor: Option<Box<dyn OutboundHttpInterceptor>>,
+}
+
+impl InstanceState {
+    pub fn set_request_interceptor(
+        &mut self,
+        interceptor: impl OutboundHttpInterceptor + 'static,
+    ) -> anyhow::Result<()> {
+        if self.request_interceptor.is_some() {
+            anyhow::bail!("set_request_interceptor can only be called once");
+        }
+        self.request_interceptor = Some(Box::new(interceptor));
+        Ok(())
+    }
 }
 
 impl SelfInstanceBuilder for InstanceState {}
+
+pub type Request = http::Request<wasmtime_wasi_http::body::HyperOutgoingBody>;
+
+/// SelfRequestOrigin indicates the base URI to use for "self" requests.
+///
+/// This is meant to be set on [`Request::extensions_mut`] in appropriate
+/// contexts such as an incoming request handler.
+#[derive(Clone, Debug)]
+pub struct SelfRequestOrigin {
+    pub scheme: Scheme,
+    pub authority: Authority,
+}
+
+impl SelfRequestOrigin {
+    pub fn from_uri(uri: &Uri) -> anyhow::Result<Self> {
+        Ok(Self {
+            scheme: uri.scheme().context("URI missing scheme")?.clone(),
+            authority: uri.authority().context("URI missing authority")?.clone(),
+        })
+    }
+
+    fn into_uri(self, path_and_query: Option<PathAndQuery>) -> Uri {
+        let mut parts = Parts::default();
+        parts.scheme = Some(self.scheme);
+        parts.authority = Some(self.authority);
+        parts.path_and_query = path_and_query;
+        Uri::from_parts(parts).unwrap()
+    }
+
+    fn use_tls(&self) -> bool {
+        self.scheme == Scheme::HTTPS
+    }
+
+    fn host_header(&self) -> HeaderValue {
+        HeaderValue::from_str(self.authority.as_str()).unwrap()
+    }
+}
+
+pub trait OutboundHttpInterceptor: Send + Sync {
+    /// Intercept an outgoing HTTP request.
+    ///
+    /// If this method returns `None`, the (possibly updated) request and config
+    /// will be passed on to the default outgoing request handler.
+    ///
+    /// If this method returns `Some(...)`, the inner result will be returned as
+    /// the result of the request, bypassing the default handler.
+    fn intercept(&self, intercepted: Intercepted)
```

Comment:
Instead of this returning an `Option` and sneaking the origin through the request extensions, I wonder if instead we could make `OutboundHttpInterceptor` generic and have:

```rust
pub trait OutboundHttpInterceptor<T>: Send + Sync {
    fn intercept(&self, intercepted: Intercepted) -> Intercepted<T>;
}

enum InterceptionResult<T> {
    Intercepted(HttpResult<HostFutureIncomingResponse>),
    NotIntercepted(T)
}
```

Then in `impl` the origin is stored in `InterceptionResult::NotIntercepted` which eliminates the implicit storing of the information in the request.

There's no need to block on this idea, we can also try it out after we merge this.

Reply:
The enum result type is probably a good idea regardless, but I personally think the extension is fine here 🤷

Reply:
It's a balance - the reason I'm not a fan of extension approach is that it's a very roundabout public API. In order to unlock support for self, users of this factor need to know that they have to somehow add a `SelfRequestOrigin` to the request extensions. It works, and we certainly can make sure this documented somewhere, but it's not a particularly ergonomic or intuitive API. 

I get that this library is unlikely to be used by a large number of people, so if we can't be super ergonomic/easy to understand, that's not the end of the world, but for my future self's sake, I'd love for us to try to make this more straight forward.

Reply:
Added return enum `InterceptOutcome`